File: src/main/java/dragonbook/inter/expr/Not.java
package dragonbook.inter.expr;

import dragonbook.lexer.Token;

public class Not extends Logical {
    public Not(Token op, Expr expr) {
        super(op, expr, expr);
    }

    @Override
    public String toString() {
        return op + " " + expr2;
    }
}

File: src/main/kotlin/dragonbook/inter/expr/Not.kt
package dragonbook.inter.expr

import dragonbook.error.ParseError
import dragonbook.lexer.Token
import dragonbook.symbols.Type

data class Not(override val op: Token, val expr: Expr) : Logical {

    init {
        validate()
    }

    override fun validate() {
        if (expr.type != Type.BOOL) throw ParseError(node = this)
    }

    override val expr1: Expr get() = expr
    override val expr2: Expr get() = expr

    override fun toString(): String = "$op $expr"
}